ert/data/_record.py

```diff
@@ -141,46 +144,70 @@ class BlobRecord(Record):
 
 class RecordCollection(_DataElement):
     records: Union[Tuple[NumericalRecord, ...], Tuple[BlobRecord, ...]]
-    ensemble_size: Optional[int] = None
+    ensemble_size: Optional[NonNegativeInt] = None
 
     @property
     def record_type(self) -> Optional[RecordType]:
         return self.records[0].record_type
 
+    @validator("records")
+    def records_validator(
+        cls, records: Union[Tuple[NumericalRecord, ...], Tuple[BlobRecord, ...]]
+    ) -> Union[Tuple[NumericalRecord, ...], Tuple[BlobRecord, ...]]:
+        assert len(records) > 0
+        record_type = records[0].record_type
+        for record in records[1:]:
+            if record.record_type != record_type:
+                raise ValueError("Ensemble records must have an invariant record type")
+        return records
+
     @validator("ensemble_size", pre=True, always=True)
     def ensemble_size_validator(
-        cls, ensemble_size: Optional[int], values: Dict[str, Any]
-    ) -> Optional[int]:
-        if ensemble_size is None and "records" in values:
-            ensemble_size = len(values["records"])
-        assert ensemble_size is not None and ensemble_size > 0
+        cls, ensemble_size: Optional[NonNegativeInt], values: Dict[str, Any]
+    ) -> Optional[NonNegativeInt]:
+        if "records" in values:
+            if ensemble_size is None:
+                ensemble_size = len(values["records"])
+            elif len(values["records"]) > 1 and ensemble_size != len(values["records"]):
+                raise ValueError(
+                    "ensemble_size incompatible with the number of records"
+                )
         return ensemble_size
 
-    @root_validator(skip_on_failure=True)
-    def ensure_consistent_ensemble(
-        cls, ensemble_record: Dict[str, Any]
-    ) -> Dict[str, Any]:
-        assert "records" in ensemble_record and "ensemble_size" in ensemble_record
-        assert len(ensemble_record["records"]) == ensemble_record["ensemble_size"]
-        record_type = ensemble_record["records"][0].record_type
-        for record in ensemble_record["records"][1:]:
-            if record.record_type != record_type:
-                raise ValueError("Ensemble records must have a uniform record type")
-        return ensemble_record
+    def __iter__(self) -> Iterator[Record]:  # type: ignore
+        assert self.ensemble_size is not None
+        if self.ensemble_size != len(self.records):
```

Comment:
I think `if self.is_invarient():` reads a bit better, but I guess is a matter of preference.